src/Crdt/Db/CrdtRepository.cs

```diff
@@ -0,0 +1,260 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.Extensions.Options;
+
+namespace Crdt.Db;
+
+public class CrdtRepository(CrdtDbContext _dbContext, IOptions<CrdtConfig> crdtConfig, DateTimeOffset? currentTime = null)
+{
+    public Task<IDbContextTransaction> BeginTransactionAsync()
+    {
+        return _dbContext.Database.BeginTransactionAsync();
+    }
+
+
+    public async Task<bool> HasCommit(Guid commitId)
+    {
+        return await _dbContext.Commits.AnyAsync(c => c.Id == commitId);
+    }
+
+    public async Task<(Commit? oldestChange, Commit[] newCommits)> FilterExistingCommits(ICollection<Commit> commits)
+    {
+        Commit? oldestChange = null;
+        var commitIdsToExclude = await _dbContext.Commits.Where(c => commits.Select(c => c.Id).Contains(c.Id))
+            .Select(c => c.Id).ToArrayAsync();
+        var newCommits = commits.ExceptBy(commitIdsToExclude, c => c.Id).Select(commit =>
+        {
+            if (oldestChange is null || commit.CompareKey.CompareTo(oldestChange.CompareKey) < 0) oldestChange = commit;
+            return commit;
+        }).ToArray(); //need to use ToArray because the select has side effects that must trigger before this method returns
+        return (oldestChange, newCommits);
+    }
+
+    public async Task DeleteStaleSnapshots(Commit oldestChange)
+    {
+        //use the oldest commit added to clear any snapshots that are based on a now incomplete history
+        await _dbContext.Snapshots
+            .Where(s => s.Commit.HybridDateTime.DateTime > oldestChange.DateTime
+                        || (s.Commit.HybridDateTime.DateTime == oldestChange.DateTime &&
+                            s.Commit.HybridDateTime.Counter > oldestChange.HybridDateTime.Counter)
+                        || (s.Commit.HybridDateTime.DateTime == oldestChange.DateTime &&
+                            s.Commit.HybridDateTime.Counter == oldestChange.HybridDateTime.Counter &&
+                            s.CommitId > oldestChange.Id)
+            )
+            .ExecuteDeleteAsync();
+    }
+
+    public IQueryable<Commit> CurrentCommits()
+    {
+        return _dbContext.Commits.DefaultOrder().Where(c => currentTime == null || c.HybridDateTime.DateTime <= currentTime);
+    }
+
+    public IQueryable<ObjectSnapshot> CurrentSnapshots()
+    {
+        return _dbContext.Snapshots.Where(snapshot => CurrentSnapshotIds().Contains(snapshot.Id));
+    }
+
+    private IQueryable<Guid> CurrentSnapshotIds()
+    {
+        return _dbContext.Snapshots.GroupBy(s => s.EntityId,
+            (entityId, snapshots) => snapshots
+                .OrderByDescending(s => s.Commit.HybridDateTime.DateTime)
+                .ThenBy(s => s.Commit.HybridDateTime.Counter)
+                .ThenBy(s => s.CommitId)
+                .First(s => currentTime == null || s.Commit.HybridDateTime.DateTime <= currentTime).Id);
```

Comment:
yeah that's certainly something that I could handle better and normalize all into one place. It can be difficult to reuse linq stuff though.

Reply:
But, to my question:
Here you've got:
```
                .OrderByDescending(s => s.Commit.HybridDateTime.DateTime)
                .ThenBy(s => s.Commit.HybridDateTime.Counter)
```
I think that would result in something like this, because the counter is sorted in ascending order:
```
- Date 1. Count 1
- Date 1. Count 2
```
Then you take `First()`: `Date 1. Count 1`
But you're looking for the newest: `Date 1. Count 2`.

Reply:
refactored